processing/src/main/java/io/druid/segment/data/CompressedVSizeIntsIndexedSupplier.java

```diff
@@ -0,0 +1,393 @@
+  private final int numBytes;
+  private final int bigEndianShift;
+  private final int littleEndianMask;
+  private final GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers;
+  private final CompressedObjectStrategy.CompressionStrategy compression;
+
+  CompressedVSizeIntsIndexedSupplier(
+      int totalSize,
+      int sizePer,
+      int numBytes,
+      GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers,
+      CompressedObjectStrategy.CompressionStrategy compression
+  )
+  {
+    Preconditions.checkArgument(
+        sizePer == (1 << Integer.numberOfTrailingZeros(sizePer)),
+        "Number of entries per chunk must be a power of 2"
+    );
+
+    this.totalSize = totalSize;
+    this.sizePer = sizePer;
+    this.baseBuffers = baseBuffers;
+    this.compression = compression;
+    this.numBytes = numBytes;
+    this.bigEndianShift = Integer.SIZE - (numBytes << 3); // numBytes * 8
+    this.littleEndianMask = (int)((1L << (numBytes << 3)) - 1); // set numBytes * 8 lower bits to 1
+  }
+
+  public static int maxIntsInBufferForBytes(int numBytes)
+  {
+    int maxSizePer = (CompressedPools.BUFFER_SIZE - bufferPadding(numBytes)) / numBytes;
+    // round down to the nearest power of 2
+    return 1 << (Integer.SIZE - 1 - Integer.numberOfLeadingZeros(maxSizePer));
+  }
+
+  private static int bufferPadding(int numBytes)
+  {
+    // when numBytes == 3 we need to pad the buffer to allow reading an extra byte
+    // beyond the end of the last value, since we use buffer.getInt() to read values.
+    // for numBytes < 3 we remove the need for padding by reading short or bytes directly.
+    if (numBytes == 3) {
+      return Ints.BYTES - numBytes;
+    } else {
+      return 0;
+    }
+  }
+
+  public static int maxIntsInBufferForValue(int maxValue)
+  {
+    return maxIntsInBufferForBytes(VSizeIndexedInts.getNumBytesForMax(maxValue));
+  }
+
+  public int size()
+  {
+    return totalSize;
+  }
+
+  @Override
+  public IndexedInts get()
+  {
+    // optimized versions for int, short, and byte columns
+    if(numBytes == Ints.BYTES) {
+      return new CompressedFullSizeIndexedInts();
+    } else if (numBytes == Shorts.BYTES) {
+      return new CompressedShortSizeIndexedInts();
+    } else if (numBytes == 1) {
+      return new CompressedByteSizeIndexedInts();
+    } else {
+      // default version of everything else, i.e. 3-bytes per value
+      return new CompressedVSizeIndexedInts();
+    }
+  }
+
+
+  public long getSerializedSize()
+  {
+    return 1 +             // version
+           1 +             // numBytes
+           Ints.BYTES + // totalSize
+           Ints.BYTES + // sizePer
+           1 +             // compression id
+           baseBuffers.getSerializedSize(); // data
+  }
+
+  public void writeToChannel(WritableByteChannel channel) throws IOException
+  {
+    channel.write(ByteBuffer.wrap(new byte[]{version, (byte) numBytes}));
+    channel.write(ByteBuffer.wrap(Ints.toByteArray(totalSize)));
```

Comment:
Same comment as other class regarding if there is proper enforcement of byte order here.